### ISSUE TITLE
Raft leader enters an election loop when a follower restarts on kubernetes

### ISSUE THREAD

**Issue by deepthidevaki** [atomix/atomix#1017]
**Expected behavior**

Killing follower should not start an election. Even if it starts, the election should converge.

**Actual behavior**

When the pod of the follower is restarted, the leader steps down to a follower and restarts the election and then keeps transitioning between candidate -> leader -> follower -> candidate ..

Observation:
When we debugged it, we observed that when a pod is deleted, its dns address cannot be resolved sometime. Hence this [Address#address(boolean)](https://github.com/atomix/atomix/blob/master/utils/src/main/java/io/atomix/utils/net/Address.java#L167-L181) can return null. When  [messaging service](https://github.com/atomix/atomix/blob/master/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java#L486) resolves the address to null, it binds the address to the wildcard `0.0.0.0` which is the behavior of `InetSocketAddress`. So the messages that the leader sends to the dead node is received by the leader itself, which causes the leader to stepdown [here](https://github.com/atomix/atomix/blob/3a94b7c80576d762dd0d396d4645df07a0b37c31/protocols/raft/src/main/java/io/atomix/protocols/raft/roles/LeaderRole.java#L549).
There might be similar problem in [NettyUnicastService#unicast(Address, String, byte[])](https://github.com/atomix/atomix/blob/3a94b7c80576d762dd0d396d4645df07a0b37c31/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyUnicastService.java#L85)

**Steps to reproduce**

Create a kubernet cluster using statefulset with replicas 3 and 
 Atomix configuration:
```
cluster {
        discovery {
            type: bootstrap
            nodes.1 {
              id: atomix-0
              address: "atomix-0.atomix.default.svc.cluster.local:5679"
            }
          }
    }
    managementGroup {
      type: raft
      partitions: 1
      storage.level: disk
      members: [atomix-1, atomix-2, atomix-0]
    }
    partitionGroups.consensus {
      type: raft
      partitions: 1
      partitionSize: 3
      storage.level: disk
      members: [atomix-1, atomix-2, atomix-0]
    }
```
Once all nodes have joined the Raft group, pick a node which is the follower for partition 1, for eg:- atomix-2. Delete pod for atomix-2 
`kubectl delete pod atomix-2`

**Environment**

 - Atomix: master (also observed in 3.1.5)
 - OS: Linux 4.14.91+ (debian base image running on kubernetes)
 - JVM: java 8

<sup>:warning:️ Please verify that your issue still occurs on the latest version of Atomix before reporting. 
The documentation is currently work-in-progress and is not yet complete.</sup>

-----------------------------------------------------------------------------------------
*Have you searched the CLOSED issues already? How about checking in with the [Atomix Google Group](https://groups.google.com/forum/#!forum/atomixio)?*

**Comment by npepinpe** [comment]
The following tests for `NettyUnicastServiceTest` will fail:

```java
@Test
  public void testUnicastUnresolvable() throws Exception {
    final AtomicBoolean received = new AtomicBoolean();
    final Address unresolvable = Address.from("unknown.local", address2.port());
    service2.addListener("test", (address, payload) -> {
      received.set(true);
      resume();
    });

    service2.unicast(unresolvable, "test", "Hello world!".getBytes());
    await(5000);
    assertFalse(received.get());
  }
```

And the following in `NettyMessagingServiceTest` will also fail:

```java
@Test
  public void testUnicastUnresolvable() {
    final Address unresolvable = Address.from("unknown.local", address1.port());
    final CountDownLatch latch1 = new CountDownLatch(1);
    final String subject = nextSubject();
    final AtomicBoolean received = new AtomicBoolean();

    final BiFunction<Address, byte[], byte[]> handler = (ep, data) -> {
      received.set(true);
      return "hello there".getBytes();
    };
    netty2.registerHandler(subject, handler, MoreExecutors.directExecutor());
    
    final CompletableFuture<Void> response = netty1.sendAsync(unresolvable, subject, "hello world".getBytes());
    response.whenComplete((r, e) -> {
      received.set(true);
      latch1.countDown();
    });
    Uninterruptibles.awaitUninterruptibly(latch1);


    assertFalse(received.get());
  }
```

In each case the sender is also receiving the response even though it tried to send to an unresolvable address.

**Comment by npepinpe** [comment]
I think this was solved by #1018
